clone! macro makes error messages cryptic

Hello,

when using the `clone!` macro rustc doesn't report the line where the errror occured correctly and instead just points to the entire macro invocation. This forces users of the macro to guess where the error occured, which isn't trivial when the function `clone!` encapsulates is longer than a few lines.

E.g. when I forget to put a semicolon somewhere the error message looks like this:

```
error: expected one of `.`, `;`, `?`, or an operator, found `self_`
   --> src/widgets/sync_list_box.rs:133:40
    |
133 |                   .connect_row_activated(clone!(@weak obj => move |list_box, row| {
    |  ________________________________________^
134 | |                     let self_ = imp::HealthSyncListBox::from_instance(&obj);
135 | |                     if (row == &self_.google_fit_start_sync_row.get()) {
136 | |                         self_.google_fit_stack.set_visible(true);
...   |
180 | |                     }
181 | |                 }));
    | |__________________^ expected one of `.`, `;`, `?`, or an operator
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

So I have to check some 50 lines to know where I forgot the semicolon.

See https://github.com/Cogitri/clone-reproducer for a reproducer:

```
error[E0599]: no method named `qiut` found for struct `gtk4::Application` in the current scope
   --> src/bin/entry_completion.rs:101:9
    |
101 | /         glib::clone!(@weak application => move |_action, _parameter| {
102 | |             application.qiut();
103 | |         }),
    | |__________^ method not found in `gtk4::Application`
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

In this case the error message seems OK because it's only a one line function, but ideally it'd point to only line 102 and not 101/103.

--- a/glib-macros/src/clone.rs
+++ b/glib-macros/src/clone.rs
@@ -2,7 +2,7 @@
 
 use crate::utils::crate_ident_new;
 use proc_macro::token_stream::IntoIter as ProcIter;
-use proc_macro::{Delimiter, Group, TokenStream, TokenTree};
+use proc_macro::{Delimiter, Group, Ident, Punct, Spacing, Span, TokenStream, TokenTree};
 use std::iter::Peekable;
 
 #[derive(Clone, Copy, Debug)]
@@ -452,13 +452,13 @@ fn check_before_closure(parts: &mut Peekable<ProcIter>) {
     }
 }
 
-fn get_closure(parts: &mut Peekable<ProcIter>) -> String {
-    let mut ret = String::new();
+fn get_closure(parts: &mut Peekable<ProcIter>) -> Vec<TokenTree> {
+    let mut ret = Vec::new();
 
     loop {
         match parts.next() {
             Some(TokenTree::Punct(p)) if p.to_string() == "|" => break,
-            Some(x) => ret.push_str(&x.to_string()),
+            Some(x) => ret.push(x),
             None => panic!("Unexpected end 3"),
         }
     }
@@ -539,22 +539,66 @@ pub(crate) fn clone_inner(item: TokenStream) -> TokenStream {
     let return_kind = parse_return_kind(&mut parts);
     check_before_closure(&mut parts);
     let closure = get_closure(&mut parts);
-    let body = tokens_to_string(parts);
+    let mut body = TokenStream::new();
+    body.extend(parts.collect::<Vec<_>>());
 
-    let x = format!(
-        "{{\n{}\nmove |{}| {{\n{}\nlet ____ret = {{ {} }};\n____ret\n}}\n}}",
-        elements
-            .iter()
-            .map(|x| x.to_str_before())
-            .collect::<Vec<_>>()
-            .join("\n"),
-        closure,
-        elements
-            .iter()
-            .map(|x| x.to_str_after(&return_kind))
-            .collect::<Vec<_>>()
-            .join("\n"),
-        body,
-    );
-    x.parse().unwrap()
+    // To prevent to lose the spans in case some errors occur in the code, we need to keep `body`!
+    //
+    // If we replaced everything that follows with a `format!`, it'd look like this:
+    //
+    // format!(
+    //     "{{\n{}\nmove |{}| {{\n{}\nlet ____ret = {{ {} }};\n____ret\n}}\n}}",
+    //     elements
+    //         .iter()
+    //         .map(|x| x.to_str_before())
+    //         .collect::<Vec<_>>()
+    //         .join("\n"),
+    //     closure,
+    //     elements
+    //         .iter()
+    //         .map(|x| x.to_str_after(&return_kind))
+    //         .collect::<Vec<_>>()
+    //         .join("\n"),
+    //     body,
+    // )
+    let mut ret: Vec<TokenTree> = vec![];
+    for el in &elements {
+        let stream: TokenStream = el
+            .to_str_before()
+            .parse()
+            .expect("failed to convert element");
+        ret.extend(stream.into_iter().collect::<Vec<_>>());
+    }
+    ret.extend(vec![
+        TokenTree::Ident(Ident::new("move", Span::call_site())),
+        TokenTree::Punct(Punct::new('|', Spacing::Alone)),
+    ]);
+    ret.extend(closure);
+    ret.extend(vec![TokenTree::Punct(Punct::new('|', Spacing::Alone))]);
+    let mut inner: Vec<TokenTree> = Vec::new();
+    for el in elements {
+        let stream: TokenStream = el
+            .to_str_after(&return_kind)
+            .parse()
+            .expect("failed to convert element after");
+        inner.extend(stream.into_iter().collect::<Vec<_>>());
+    }
+    // The commented lines that follow *might* be useful, don't know. Just in case, I'm keeping
+    // them around. You're welcome future me!
+    inner.extend(vec![
+        // TokenTree::Ident(Ident::new("let", Span::call_site())),
+        // TokenTree::Ident(Ident::new("____ret", Span::call_site())),
+        // TokenTree::Punct(Punct::new('=', Spacing::Alone)),
+        TokenTree::Group(Group::new(Delimiter::Brace, body)),
+        // TokenTree::Punct(Punct::new(';', Spacing::Alone)),
+        // TokenTree::Ident(Ident::new("____ret", Span::call_site())),
+    ]);
+    let mut inners = TokenStream::new();
+    inners.extend(inner);
+    ret.extend(vec![TokenTree::Group(Group::new(Delimiter::Brace, inners))]);
+
+    let mut rets = TokenStream::new();
+    rets.extend(ret);
+
+    TokenTree::Group(Group::new(Delimiter::Brace, rets)).into()
 }

--- a/glib/tests/clone.rs
+++ b/glib/tests/clone.rs
@@ -400,11 +400,15 @@ fn test_clone_macro_double_rename() {
     closure(0i8); // to prevent compiler error for unknown `x` type.
     let _ = clone!(@weak v, @weak w as _x => @default-return true, move || false);
 
-    let closure = clone!(@strong v as _x, @strong w => @default-return true, move |_| { false });
+    let closure = clone!(@strong v as _x, @strong w => @default-return true, move |_| {
+        false
+    });
     closure(0i8); // to prevent compiler error for unknown `x` type.
     let _ = clone!(@strong v as _x, @strong w => @default-return true, move || false);
 
-    let closure = clone!(@strong v, @strong w as _x => @default-return true, move |_| { false });
+    let closure = clone!(@strong v, @strong w as _x => @default-return true, move |_| {
+        false
+    });
     closure(0i8); // to prevent compiler error for unknown `x` type.
     let _ = clone!(@strong v, @strong w as _x => @default-return true, move || false);
 }
@@ -450,14 +454,14 @@ fn test_clone_macro_default_return() {
 
     struct Foo(i32);
 
-    let _closure = clone!(@weak v => @default-return Foo(0), move || { Foo(1) });
+    let _closure = clone!(@weak v => @default-return Foo(0), move || Foo(1));
 
     #[allow(dead_code)]
     struct Bar {
         x: i32,
     }
 
-    let _closure = clone!(@weak v => @default-return Bar { x: 0 }, move || { Bar { x: 1 } });
+    let _closure = clone!(@weak v => @default-return Bar { x: 0 }, move || Bar { x: 1 });
 
     #[allow(dead_code)]
     enum Enum {
@@ -465,10 +469,10 @@ fn test_clone_macro_default_return() {
         B(i32),
         C { x: i32 },
     }
-    let _closure = clone!(@weak v => @default-return Enum::A, move || { Enum::A });
-    let _closure = clone!(@weak v => @default-return Enum::B(0), move || { Enum::A });
-    let _closure = clone!(@weak v => @default-return Enum::C { x: 0 }, move || { Enum::A });
-    let _closure = clone!(@weak v => @default-return { let x = 12; x + 2 }, move || { 19 });
+    let _closure = clone!(@weak v => @default-return Enum::A, move || Enum::A);
+    let _closure = clone!(@weak v => @default-return Enum::B(0), move || Enum::A);
+    let _closure = clone!(@weak v => @default-return Enum::C { x: 0 }, move || Enum::A);
+    let _closure = clone!(@weak v => @default-return { let x = 12; x + 2 }, move || 19);
 }
 
 #[test]

